File: sdk/merklith-sdk-ts/src/contract.ts
import { MerklithClient } from './client';
import { Address } from './types';

export interface ContractOptions {
  address: Address;
  abi?: any[];
}

export interface CallOptions {
  from?: Address;
  gas?: bigint;
  gasPrice?: bigint;
  value?: bigint;
}

export interface SendOptions extends CallOptions {
  nonce?: number;
}

export class Contract {
  private client: MerklithClient;
  public address: Address;
  private abi?: any[];

  constructor(client: MerklithClient, options: ContractOptions) {
    this.client = client;
    this.address = options.address;
    this.abi = options.abi;
  }

  async call(data: string, options: CallOptions = {}): Promise<string> {
    return this.client.call({
      to: this.address,
      data,
      from: options.from,
      gas: options.gas ? `0x${options.gas.toString(16)}` : undefined,
      gasPrice: options.gasPrice ? `0x${options.gasPrice.toString(16)}` : undefined,
      value: options.value ? `0x${options.value.toString(16)}` : undefined,
    });
  }

  async sendTransaction(data: string, options: SendOptions = {}): Promise<string> {
    // This would need a wallet to sign the transaction
    // For now, just return a placeholder
    throw new Error('Send transaction requires wallet integration');
  }

  async getCode(): Promise<string> {
    return this.client.getCode(this.address);
  }

  async exists(): Promise<boolean> {
    const code = await this.getCode();
    return code !== '0x' && code.length > 2;
  }
}

export class ContractDeployer {
  private client: MerklithClient;
  private bytecode: string;
  private abi?: any[];

  constructor(client: MerklithClient, bytecode: string, abi?: any[]) {
    this.client = client;
    this.bytecode = bytecode;
    this.abi = abi;
  }

  async deploy(args: string = '0x', options: SendOptions = {}): Promise<Contract> {
    // Combine bytecode with constructor args
    const data = this.bytecode + args.replace('0x', '');

    // This would need wallet integration
    throw new Error('Contract deployment requires wallet integration');
  }
}

File: sdk/merklith-sdk-rs/src/contract.rs
//! Contract interaction helpers.

use merklith_types::{Address, Transaction, U256};
use std::marker::PhantomData;

use crate::client::Client;
use crate::errors::{Result, SdkError};
use crate::types::{CallOptions, TxOptions};

/// Contract interface.
pub struct Contract {
    client: Client,
    address: Address,
    abi: Option<serde_json::Value>,
}

impl Contract {
    /// Create new contract interface.
    pub fn new(client: Client, address: Address) -> Self {
        Self {
            client,
            address,
            abi: None,
        }
    }

    /// Create with ABI.
    pub fn with_abi(client: Client, address: Address, abi: serde_json::Value) -> Self {
        Self {
            client,
            address,
            abi: Some(abi),
        }
    }

    /// Get contract address.
    pub fn address(&self) -> Address {
        self.address
    }

    /// Call a contract method (read-only).
    pub async fn call(
        &self,
        data: Vec<u8>,
        options: CallOptions,
    ) -> Result<Vec<u8>> {
        let tx = Transaction::new(
            1337, // Default chain ID
            0,
            Some(self.address),
            options.value.unwrap_or(U256::ZERO),
            options.gas_limit.unwrap_or(100_000),
            U256::from(1_000_000_000u64),
            U256::from(1_000_000_000u64),
        );

        self.client.call(&tx, Some(options.block)).await
    }

    /// Send a transaction to contract.
    pub async fn send(
        &self,
        data: Vec<u8>,
        options: TxOptions,
    ) -> Result<merklith_types::Hash> {
        let tx = Transaction::new(
            1337, // Default chain ID
            options.nonce.unwrap_or(0),
            Some(self.address),
            options.value.unwrap_or(U256::ZERO),
            options.gas_limit.unwrap_or(100_000),
            options.gas_price.unwrap_or(U256::from(1_000_000_000u64)),
            U256::from(1_000_000_000u64),
        );

        self.client.send_transaction(&tx).await
    }

    /// Get contract bytecode.
    pub async fn code(&self) -> Result<Vec<u8>> {
        self.client.get_code(&self.address).await
    }

    /// Check if contract exists.
    pub async fn exists(&self) -> Result<bool> {
        let code = self.code().await?;
        Ok(!code.is_empty())
    }
}

/// Contract builder for deployment.
pub struct ContractBuilder {
    bytecode: Vec<u8>,
    abi: Option<serde_json::Value>,
    args: Vec<u8>,
}

impl ContractBuilder {
    /// Create new builder with bytecode.
    pub fn new(bytecode: Vec<u8>) -> Self {
        Self {
            bytecode,
            abi: None,
            args: vec![],
        }
    }

    /// Set ABI.
    pub fn with_abi(mut self, abi: serde_json::Value) -> Self {
        self.abi = Some(abi);
        self
    }

    /// Set constructor arguments.
    pub fn with_args(mut self, args: Vec<u8>) -> Self {
        self.args = args;
        self
    }

    /// Deploy the contract.
    pub async fn deploy(
        self,
        client: &Client,
        options: TxOptions,
    ) -> Result<Contract> {
        // Combine bytecode with constructor args
        let mut data = self.bytecode;
        data.extend_from_slice(&self.args);

        let tx = Transaction::new(
            1337, // Default chain ID
            options.nonce.unwrap_or(0),
            None, // Contract creation
            options.value.unwrap_or(U256::ZERO),
            options.gas_limit.unwrap_or(1_000_000),
            options.gas_price.unwrap_or(U256::from(1_000_000_000u64)),
            U256::from(1_000_000_000u64),
        );

        // Send transaction
        let hash = client.send_transaction(&tx).await?;

        // Wait for receipt to get contract address
        let receipt = client.wait_for_transaction(
            &hash,
            std::time::Duration::from_secs(60),
        ).await?;

        // Get contract address from receipt
        // For now, return a placeholder
        let contract_address = Address::ZERO;

        Ok(Contract {
            client: client.clone(),
            address: contract_address,
            abi: self.abi,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_contract_creation() {
        let client = Client::new("http://localhost:8545");
        let contract = Contract::new(client, Address::ZERO);
        
        assert_eq!(contract.address(), Address::ZERO);
    }

    #[test]
    fn test_contract_builder() {
        let builder = ContractBuilder::new(vec![0x60, 0x80, 0x60])
            .with_args(vec![0x01, 0x02]);

        // Just ensure it compiles
        assert!(!builder.bytecode.is_empty());
    }
}
